### DISCUSSION TITLE
What is the best way to implement firebase analytics in Next.js

### DISCUSSION THREAD

**Discussion by udithv** [vercel/next.js#13654] [Help]
I am implementing a simple website using next.js. how can one implement firebase analytics and possibly admob in Next.js app .

Should I create a HOC component.
or 

Should I initialize it in _document.js or _app.js file

**Comment by qathom** [comment]
Hi @udithv 

I was looking for a solution and I finally created a Provider with React Context API.
For my case, I need to track page changes and push `page_view`events to Firebase.
As you might know, the Firebase analytics module must be loaded on client side which can be confusing on Nextjs.

If it helps you, here is my implementation:

**FirebaseTrackingProvider.tsx : provider + context**
```ts
import { createContext, useState, useEffect, ReactNode } from 'react';
import firebase from 'firebase/app';
import { useRouter } from 'next/router';
import 'firebase/analytics';
import config from '~/firebase/config';

if (firebase.apps.length === 0) {
  firebase.initializeApp(config);
}

export const FirebaseContext = createContext(null);

export const FirebaseTrackingProvider = (props: { children: ReactNode }) => {
  const router = useRouter();
  const [tracking, setTracking] = useState<firebase.analytics.Analytics | null>(null);

  useEffect(() => {
    setTracking(firebase.analytics());

    const handleRouteChange = (url: string) => {
      if (!tracking) {
        return;
      }

      tracking.logEvent('page_view', {
        page_location: url,
        page_title: document?.title,
      });
    };

    router.events.on('routeChangeStart', handleRouteChange);

    return () => {
      router.events.off('routeChangeStart', handleRouteChange);
    };
  }, [tracking]);

  return (
    <FirebaseContext.Provider value={tracking}>
      {props.children}
    </FirebaseContext.Provider>
  );
};
```

**Usage:**

By default, the provider is listening to route changes and will push events.
It is possible to use the context (useContext...) if we want to publish other events (push click events, custom events.).
In this way, it brings a bit of flexibility.

**The layout component with the provider**:

```ts
import { FunctionComponent, ReactNode } from 'react';

type Props = {
  children: ReactNode;
};

const Layout: FunctionComponent<Props> = ({ children }) => {
  return (
    <FirebaseTrackingProvider>
      <div className="layout">
        {children}
      </div>
    </FirebaseTrackingProvider>
  );
};

export default Layout;
```
